service/tcp_test.go

```diff
@@ -209,11 +250,220 @@ func (m *probeTestMetrics) AddUDPPacketFromTarget(clientLocation, accessKey, sta
+	}
+	s.GracefulStop()
+	require.Equal(t, len(buf), len(testMetrics.probeData))
+}
+
+func makeClientBytesBasic(t *testing.T, cipher *ss.Cipher, targetAddr string) []byte {
+	var buffer bytes.Buffer
+	socksTargetAddr := socks.ParseAddr(targetAddr)
+	require.Equal(t, 1+4+2, len(socksTargetAddr))
+	ssw := ss.NewShadowsocksWriter(&buffer, cipher)
+	n, err := ssw.Write(socksTargetAddr)
+	require.Nil(t, err, "Write failed: %v", err)
+	require.Equal(t, len(socksTargetAddr), n, "Write failed: %v", err)
+	require.Equal(t, 73, buffer.Len())
+
+	n, err = ssw.Write([]byte("initial data"))
+	require.Nil(t, err, "Write failed: %v", err)
+	require.Equal(t, 73+2+16+12+16, buffer.Len()) // 119
+
+	n, err = ssw.Write([]byte("more data"))
+	require.Nil(t, err, "Write failed: %v", err)
+	require.Equal(t, 119+2+16+9+16, buffer.Len()) // 162
+
+	return buffer.Bytes()
+}
+
+func makeClientBytesCoalesced(t *testing.T, cipher *ss.Cipher, targetAddr string) []byte {
+	var buffer bytes.Buffer
+	socksTargetAddr := socks.ParseAddr(targetAddr)
+	ssw := ss.NewShadowsocksWriter(&buffer, cipher)
+	n, err := ssw.LazyWrite(socksTargetAddr)
+	require.Nil(t, err, "LazyWrite failed: %v", err)
+	require.Equal(t, len(socksTargetAddr), n, "LazyWrite failed: %v", err)
+	n, err = ssw.Write([]byte("initial data"))
+	require.Nil(t, err, "Write failed: %v", err)
+	require.Equal(t, 32+2+16+7+12+16, buffer.Len()) // 85
+
+	n, err = ssw.Write([]byte("more data"))
+	require.Nil(t, err, "Write failed: %v", err)
+	return buffer.Bytes()
+}
+
+func firstCipher(cipherList CipherList) *ss.Cipher {
+	snapshot := cipherList.SnapshotForClientIP(nil)
+	cipherEntry := snapshot[0].Value.(*CipherEntry)
+	return cipherEntry.Cipher
+}
+
+func TestProbeClientBytesBasicTruncated(t *testing.T) {
+	listener := makeLocalhostListener(t)
+	cipherList, err := MakeTestCiphers(ss.MakeTestSecrets(1))
+	cipher := firstCipher(cipherList)
+	require.Nil(t, err, "MakeTestCiphers failed: %v", err)
+	testMetrics := &probeTestMetrics{}
+	s := NewTCPService(cipherList, nil, testMetrics, 200*time.Millisecond)
+	s.SetTargetIPValidator(allowAll)
+	go s.Serve(listener)
+
+	discardListener, discardWait := startDiscardServer(t)
+	initialBytes := makeClientBytesBasic(t, cipher, discardListener.Addr().String())
+	for numBytesToSend := 0; numBytesToSend < len(initialBytes); numBytesToSend++ {
+		t.Logf("Sending %v bytes", numBytesToSend)
+		bytesToSend := initialBytes[:numBytesToSend]
+		err := probe(listener.Addr().(*net.TCPAddr), bytesToSend)
+		require.Nil(t, err, "Failed for %v bytes sent: %v", numBytesToSend, err)
+	}
+	s.GracefulStop()
+	statusCount := testMetrics.countStatuses()
+	require.Equal(t, 50, statusCount["ERR_CIPHER"])
+	require.Equal(t, 7+16, statusCount["ERR_READ_ADDRESS"])
+	require.Equal(t, 2, statusCount["OK"]) // On the chunk boundaries.
+	require.Equal(t, len(initialBytes)-50-7-16-2, statusCount["ERR_RELAY_CLIENT"])
+	// We only count as probes failures in the first 50 bytes.
```

Comment:
This may be true, but does this still make sense?  Maybe we should report decryption errors similarly regardless of how many bytes were sent.